modules/blockbuilder/tenant_store_test.go

```diff
@@ -92,3 +98,113 @@ func TestAdjustTimeRangeForSlack(t *testing.T) {
 		})
 	}
 }
+
+func TestTenantStoreEndToEndHistoricalData(t *testing.T) {
+	var (
+		count     = 3
+		startTime = time.Now().Add(-24 * time.Hour)
+	)
+
+	testCases := []struct {
+		name              string
+		traceStart        time.Time // All spans will start at this time
+		traceEnd          time.Time // All spans will end at this time
+		cycleDuration     time.Duration
+		slackDuration     time.Duration
+		expectedStartTime time.Time
+		expectedEndTime   time.Time
+	}{
+		{
+			name:              "all good",
+			traceStart:        startTime,
+			traceEnd:          startTime.Add(time.Minute),
+			cycleDuration:     5 * time.Minute,
+			slackDuration:     5 * time.Minute,
+			expectedStartTime: startTime,
+			expectedEndTime:   startTime.Add(time.Minute),
+		},
+		{
+			name:              "before start",
+			traceStart:        startTime.Add(-10 * time.Minute),
+			traceEnd:          startTime,
+			cycleDuration:     5 * time.Minute,
+			slackDuration:     5 * time.Minute,
+			expectedStartTime: startTime.Add(-5 * time.Minute), // Only goes as early as -slack
+			expectedEndTime:   startTime,
+		},
+		{
+			name:              "after end",
+			traceStart:        startTime,
+			traceEnd:          startTime.Add(20 * time.Minute),
+			cycleDuration:     5 * time.Minute,
+			slackDuration:     5 * time.Minute,
+			expectedStartTime: startTime,
+			expectedEndTime:   startTime.Add(10 * time.Minute), // Only goes as far as cycle + slack
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			meta := writeHistoricalData(t, count, startTime, tc.cycleDuration, tc.slackDuration, tc.traceStart, tc.traceEnd)
+
+			// NOTE - Roundtripped times from the meta json don't include the monotonic clock value,
+			//        therefore we do the same on the test values by calling Truncate(0).
+			require.Equal(t, tc.expectedStartTime.Truncate(0).UTC(), meta.StartTime.UTC())
+			require.Equal(t, tc.expectedEndTime.Truncate(0).UTC(), meta.EndTime.UTC())
+
+			// Verify other properties of the block
+			require.EqualValues(t, count, meta.TotalObjects)
+			require.EqualValues(t, 1, meta.TotalRecords)
+			require.Greater(t, meta.Size_, uint64(0))
+		})
+	}
+}
+
+func writeHistoricalData(t *testing.T, count int, startTime time.Time, cycleDuration, slackDuration time.Duration, traceStart, traceEnd time.Time) *backend.BlockMeta {
```

Comment:
Maybe you can reuse this:

https://github.com/grafana/tempo/blob/1b998e9437031f82888f95025f586ad28b7f3a37/modules/blockbuilder/blockbuilder_test.go#L562